streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java

```diff
@@ -128,47 +128,101 @@ public void process(final Record<KLeft, Change<VLeft>> record) {
         }
 
         private void leftJoinInstructions(final Record<KLeft, Change<VLeft>> record) {
```

Comment:
The diff is unfortunately a little bit messy, but if you compare new and old code, also because I added comments, it should become clear the the new code is much cleaner, especially for left-join case.